File: System-Design/Caching/LRU-Cache/lru_cache.py
from collections import OrderedDict

class LRUCache:
    def __init__(self, capacity: int):
        self.cache = OrderedDict()
        self.capacity = capacity

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        self.cache.move_to_end(key)
        return self.cache[key]

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache.move_to_end(key)
        self.cache[key] = value
        if len(self.cache) > self.capacity:
            self.cache.popitem(last=False) 

if __name__ == '__main__':
    lru = LRUCache(2)
    lru.put(1, 10)
    lru.put(2, 20)
    assert lru.get(1) == 10  
    lru.put(3, 30)          
    assert lru.get(2) == -1 
    
    print("Python LRU Cache Test Passed!")

File: System-Design/Caching/LRU-Cache/lru_cache.cpp
#include <iostream>
#include <unordered_map>
#include <list>
#include <cassert>

class LRUCache {
private:
    int capacity;
    std::list<std::pair<int, int>> cacheList;
    std::unordered_map<int, std::list<std::pair<int, int>>::iterator> cacheMap;

public:
    LRUCache(int cap) : capacity(cap) {}

    int get(int key) {
        if (cacheMap.find(key) == cacheMap.end()) return -1;
        cacheList.splice(cacheList.begin(), cacheList, cacheMap[key]);
        return cacheMap[key]->second;
    }

    void put(int key, int value) {
        if (cacheMap.find(key) != cacheMap.end()) {
            cacheList.splice(cacheList.begin(), cacheList, cacheMap[key]);
            cacheMap[key]->second = value;
            return;
        }
        if (cacheList.size() == capacity) {
            auto last = cacheList.back();
            cacheMap.erase(last.first);
            cacheList.pop_back();
        }
        cacheList.push_front({key, value});
        cacheMap[key] = cacheList.begin();
    }
};

int main() {
    LRUCache lru(2);
    lru.put(1, 10);
    lru.put(2, 20);
    assert(lru.get(1) == 10); 
    lru.put(3, 30);          
    assert(lru.get(2) == -1);
    
    std::cout << "C++ LRU Cache Test Passed!\n";
    return 0;
}